source/particle/generator/interface.cc

```diff
@@ -0,0 +1,191 @@
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      std::pair<types::LevelInd,Particle<dim> >
+      Interface<dim>::generate_particle(const Point<dim> &position,
+                                        const types::particle_index id) const
+      {
+        const typename parallel::distributed::Triangulation<dim>::active_cell_iterator it =
+          (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), position)).first;;
+
+        //Only try to add the point if the cell it is in, is on this processor
+        if (it->is_locally_owned())
+          {
+            const Particle<dim> particle(position, id);
+            const types::LevelInd cell(it->level(), it->index());
+            return std::make_pair(cell,particle);
+          }
+
+        AssertThrow(false,
+                    ExcMessage("A particle generator plugin tried to create a particle in"
+                               "a location that is not owned by the current processor. Please "
+                               "check that your particle generator ensures correct locations "
+                               "for the generated particles."));
+
+        // Avoid warnings about missing return
+        const Particle<dim> particle;
+        const types::LevelInd cell(0,0);
+        return std::make_pair(cell,particle);
+      }
+
+      template <int dim>
+      void
+      Interface<dim>::declare_parameters (ParameterHandler &)
+      {}
+
+      template <int dim>
+      void
+      Interface<dim>::parse_parameters (ParameterHandler &)
+      {}
+
+
+// -------------------------------- Deal with registering models and automating
+// -------------------------------- their setup and selection at run time
+
+      namespace
+      {
+        std_cxx1x::tuple
+        <void *,
+        void *,
+        internal::Plugins::PluginList<Interface<2> >,
+        internal::Plugins::PluginList<Interface<3> > > registered_plugins;
+      }
+
+
+
+      template <int dim>
+      void
+      register_particle_generator (const std::string &name,
+                                   const std::string &description,
+                                   void (*declare_parameters_function) (ParameterHandler &),
+                                   Interface<dim> *(*factory_function) ())
+      {
+        std_cxx1x::get<dim>(registered_plugins).register_plugin (name,
+                                                                 description,
+                                                                 declare_parameters_function,
+                                                                 factory_function);
+      }
+
+
+      template <int dim>
+      Interface<dim> *
+      create_particle_generator (ParameterHandler &prm)
+      {
+        std::string name;
+        prm.enter_subsection ("Postprocess");
+        {
+          prm.enter_subsection ("Tracers");
```

Comment:
Yes, I think so. The only "postprocessor-like" thing about them, is the tracer output (which writes them into a file, this was fine, when this was all, we could do with them). But the postprocessor status is limiting us as soon as we switch to active tracers (e.g. in case of nonlinear iterations, postprocessors are only called after all iterations have finished). I could imagine having an input section "Particle", which would also follow our folder structure. We would eventually have to decide, which plugins belong into which section. "output" should continue to be in the "Postprocess"->"Tracers" section, but the others might well go into "Particle".